### PR TITLE
Prepare for phpstan level 4

### PR DESCRIPTION
**Pull request by dereuromark**
- [x] Requires https://github.com/propelorm/Propel2/pull/1607 to be merged first.

~~Down from 140 issues to 99 so far.~~
~~<78 issues left~~
41 issues left

All open issues including level 5:
61 errors
These might have to go into baseline for now

### CODE REVIEW

File: src/Propel/Generator/Behavior/Archivable/templates/objectRestoreFromArchive.php [hunk 1]
```diff
@@ -7,7 +7,7 @@
  *
  * @throws PropelException If the object has no corresponding archive.
  *
- * @return $this|<?php echo $objectClassName ?> The current object (for fluent API support)
+ * @return $this The current object (for fluent API support)
```
[the remark this answers is not in the document]
Comment (dereuromark) [reply src/Propel/Generator/Behavior/Archivable/templates/objectRestoreFromArchive.php:10]: The generated code must/can return only $this, so I don't think we need to ignore this if we can fix it here.

File: src/Propel/Generator/Behavior/Sluggable/SluggableBehavior.php [hunk 2]
```diff
@@ -63,15 +63,15 @@ public function modifyTable()
...
+
     /**
      * Adds a unique constraint to the table to enforce uniqueness of the slug_column
      *
      * @param Table $table
      */
     protected function addUniqueConstraint(Table $table)
     {
-        $unique = new Unique($this->getColumnForParameter('slug_column'));
+        $unique = new Unique(); //$this->getColumnForParameter('slug_column')
```
[the remark this answers is not in the document]
Comment (dereuromark) [reply src/Propel/Generator/Behavior/Sluggable/SluggableBehavior.php:74]: So such accepted arg, this has to be set in the next line as setter

File: src/Propel/Common/Config/ConfigurationManager.php [hunk 3]
```diff
@@ -133,21 +133,21 @@ public function getConnectionParametersArray($section = 'runtime')
      * This method also looks for a '.dist' configuration file and loads it.
      *
      * @param string $fileName  Configuration file name or directory in which resides the configuration file.
```
[the remark this answers is not in the document]
Comment (dereuromark) [reply src/Propel/Common/Config/ConfigurationManager.php:135]: I dont think this makes sense for a load() call that requires an input arg.

[the remark this answers is not in the document]
Comment (demkos) [reply src/Propel/Common/Config/ConfigurationManager.php:135]: But we will have to fix it later, for next increasing of phpstan level, as `$filename` comes from `__constructor` where 
```
* @param string|null $filename  Configuration file name or directory in which resides the configuration file.
```

File: src/Propel/Generator/Behavior/Sluggable/SluggableBehavior.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (demkos) [reply src/Propel/Generator/Behavior/Sluggable/SluggableBehavior.php:74]: Can we remove commented code?

File: src/Propel/Common/Config/ConfigurationManager.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (dereuromark) [reply src/Propel/Common/Config/ConfigurationManager.php:135]: I plan to also get rid of those there, where they don't make sense and instead either throw a meaningful exception or check null and not enter such code pieces where null handling is not useful.
This here would be a step back IMO until then.

[the remark this answers is not in the document]
Comment (dereuromark) [reply src/Propel/Common/Config/ConfigurationManager.php:135]: Actually, we might indeed go this way for now: https://github.com/propelorm/Propel2/pull/1628/files#diff-51e0a603c43d7a1f796b00cffaa97bb0R40 with level 5 then.
Either way this is then taken care off with follow up PRs.

File: src/Propel/Common/Config/Loader/FileLoader.php [hunk 4]
```diff
@@ -124,20 +124,20 @@ protected function checkSupports($ext, $resource)
             return ($ext === $extension);
         }
 
-        if (is_array($ext)) {
-            $supported = false;
+        if (!is_array($ext)) {
+            throw new \InvalidArgumentException('$ext must string or string[]');
```
Comment (gechetspr) [inline src/Propel/Common/Config/Loader/FileLoader.php:128]:
```suggestion
            throw new \InvalidArgumentException('$ext must be string or string[]');
```

File: src/Propel/Generator/Model/Database.php [hunk 5]
```diff
@@ -480,6 +480,33 @@ public function getTableByPhpName($phpName)
...
+     * @param Table[] $tables An array of Table instances
+     */
+    public function addTables(array $tables)
+    {
+        foreach ($tables as $table) {
+            $this->addTable($table);
+        }
+    }
+
+    public function removeTable(Table $table)
```
Comment (gechetspr) [inline src/Propel/Generator/Model/Database.php:495]: please add a doc block

File: src/Propel/Runtime/ActiveQuery/Criterion/AbstractCriterion.php [hunk 6]
```diff
@@ -321,11 +321,12 @@ public function equals($obj)
         /** @var AbstractCriterion $crit */
         $crit = $obj;
 
-        $isEquiv = (((null === $this->table && null === $crit->getTable())
+        $isEquiv = (
+            ((null === $this->table && null === $crit->getTable())
```
Comment (gechetspr) [inline src/Propel/Runtime/ActiveQuery/Criterion/AbstractCriterion.php:325]: Do we keep yoda style for now?

File: src/Propel/Runtime/Collection/OnDemandCollection.php [hunk 7]
```diff
@@ -30,7 +30,7 @@ class OnDemandCollection extends Collection
     private $lastIterator;
 
     /**
-     * @param AbstractFormatter    $formatter
+     * @param \Propel\Runtime\Formatter\ObjectFormatter    $formatter
```
Comment (gechetspr) [inline src/Propel/Runtime/Collection/OnDemandCollection.php:33]: Why not AbstractFormatter?

File: src/Propel/Runtime/ActiveQuery/Criterion/AbstractCriterion.php (hunk 6 above)
Comment (dereuromark) [reply src/Propel/Runtime/ActiveQuery/Criterion/AbstractCriterion.php:325]: jep, code sniffer will take care here later next week

File: src/Propel/Runtime/Collection/OnDemandCollection.php (hunk 7 above)
Comment (dereuromark) [reply src/Propel/Runtime/Collection/OnDemandCollection.php:33]: it doesnt have the method we need.
and parent is already the same.
refactoring the constructor would be BC break for now.

**Pull request merged**